src/high-level/abstract-tensor.lisp

```diff
@@ -0,0 +1,235 @@
+;;;; abstract-tensor.lisp
+;;;;
+;;;; Author: Cole Scott
+;;;;
+;;;; Collaborators: Robert Smith
+
+(in-package #:magicl)
+
+(defclass abstract-tensor ()
+  ()
+  (:documentation "Abstract tensor class as super class of tensor classes")
+  (:metaclass abstract-class:abstract-class))
+
+(defgeneric specialize-tensor (abstract-tensor)
+  (:documentation "Specialize a tensor to the most specfic applicable tensor class"))
+
+(defgeneric generalize-tensor (abstract-tensor)
+  (:documentation "Generalize a tensor to the least specific applicable tensor class"))
+
+;;; abstract-tensor protocol
+
+;; Must be implemented by subclasses
+
+(defgeneric shape (abstract-tensor)
+  (:documentation "The shape (dimensions) of the tensor. eg. '(2 3) for a 2x3 tensor"))
+
+(defgeneric tref (abstract-tensor &rest pos)
+  (:documentation "Get a reference to element at position"))
+
+(defgeneric (setf tref) (new-value abstract-tensor &rest pos)
+  (:documentation "Set the value of element at position"))
+
+(defgeneric copy-tensor (abstract-tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, not initializing the storage"))
+
+(defgeneric deep-copy-tensor (abstract-tensor &rest args)
+  (:documentation "Create a new tensor with the same properties as the given tensor, copying storage"))
+
+;; Stuff with somewhat sensible defaults
+
+(defgeneric rank (abstract-tensor)
```

Comment:
It's not a good idea to call this the rank of a tensor, it should be called `order` (or `valence`) instead.

Also, IMHO this should be stored as a slot within the tensor definition (O(1) operation) if shape is a list (in which case it is O(n)).

Reply:
This is named RANK to follow how common lisp does ARRAY-RANK. Order is taken by the column/row order but I am willing to move things around if it makes using magicl more intuitive

Reply:
I'll leave it up to you and others to decide whether to change it or not. My opinion is that if we are going to diverge from MATLAB/NumPy [1,2], it ought to be for a good reason and I don't think this is one. On the other hand, both Tensorflow and PyTorch adhere to this misnomer.

Similarly, I'd rename `order` to `layout`, which is the name that CBLAS uses (see [3] and [4]).

Just for the record and, again, I expect other reviewers and future users to chime in with their preferences, I'm also against calling these objects tensors but I'm not particularly invested in changing this term (I guess the name has been so widely misused, that it is now OK). They are n-dimensional arrays (NumPy gets this right and calls them NDArrays).

[1] https://www.mathworks.com/help/matlab/ref/rank.html
[2] https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.matrix_rank.html
[3] http://www.netlib.org/lapack/explore-html/de/da0/cblas_8h_a79f58380f0c81f707d29e2e85b81b794.html
[4] https://software.intel.com/en-us/mkl-developer-reference-c-matrix-storage-schemes-for-blas-routines

Reply:
See https://github.com/rigetti/magicl/pull/63#discussion_r362908519